### PR TITLE
Implement Rate Limiter Contract

### PR DESCRIPTION
**Pull request by 0xmovses**
# Summary
For a full spec please read [MIP-56](https://github.com/movementlabsxyz/MIP/pull/56)

# Changelog
- Creates `RateLimiter.sol`
- Integrates `RateLimiter` into Initiator and Counterparty contracts

# Testing
Adds the solidity unit test `testRateLimitExceeded` for both Initiator and Counterparty Contracts
`cd protocol-units/bridge/contracts && forge test`

# Outstanding issues
Integration tests with the relayer should be added in a separate PR, more solidity unit tests could be added to test out the RateLimiter.sol

### CODE REVIEW

File: protocol-units/bridge/contracts/src/RateLimiter.sol [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    // Update the security fund and recalculate rate limits
+    function updateSecurityFund(uint256 newSecurityFund) external onlyOwner {
+        securityFund = newSecurityFund;
+        _updateRateLimits();
+        emit SecurityFundUpdated(newSecurityFund);
+    }
+
+    // Private function to update the rate limits based on the security fund and risk period
+    function _updateRateLimits() private {
```
Comment (l-monninger) [inline protocol-units/bridge/contracts/src/RateLimiter.sol:74]: Looks correct, but choice of representation here and in MIP is funny. I think the MIP should express this more clearly, i.e., use $\frac{\text{securityFund}}{2 * \text{riskPeriod}}$

File: protocol-units/bridge/contracts/test/AtomicBridgeCounterpartyMOVE.t.sol [hunk 2]
```diff
@@ -51,44 +53,53 @@ contract AtomicBridgeCounterpartyMOVETest is Test {
 
         originator = vm.addr(uint256(keccak256(abi.encodePacked(block.timestamp, block.prevrandao))));
 
-        // Deploy the AtomicBridgeInitiator contract with a 48-hour time lock
+        // Deploy and initialize the RateLimiter contract
+        rateLimiter = new RateLimiter();
+        uint256 riskPeriod = 24 * 60 * 60; // 24 hours in seconds
+        uint256 securityFund = 5 ether;
```
Comment (l-monninger) [inline protocol-units/bridge/contracts/test/AtomicBridgeCounterpartyMOVE.t.sol:59]: Probably want to be clear about MOVE token units here and in runbooks.

File: protocol-units/bridge/contracts/test/AtomicBridgeCounterpartyMOVE.t.sol [hunk 3]
```diff
@@ -180,89 +191,152 @@ contract AtomicBridgeCounterpartyMOVETest is Test {
...
+
+        // Attempt a second transfer to exceed the rate limit
+        bytes32 secondBridgeTransferId = keccak256(
+            abi.encodePacked(block.timestamp, initiator, recipient, moveAmount, hashLock, timeLock)
+        );
+
+        // Expect the second transfer to exceed the rate limit and revert
+        vm.expectRevert("RATE_LIMIT_EXCEEDED");
+        atomicBridgeCounterpartyMOVE.lockBridgeTransfer(
+            initiator,
```
Comment (l-monninger) [inline protocol-units/bridge/contracts/test/AtomicBridgeCounterpartyMOVE.t.sol:333]: 1. I'd like to see tests demonstrating both the L1->L2 direction and the L2->L1 direction.
2. For user experience, L1->L2 should be able to revert on `initiateBridgeTransfer`, no?
3. In general, why don't we use the `bridgeTransferId` returned by initiate?

**Review by l-monninger — CHANGES_REQUESTED** [review]

**Comment by l-monninger** [comment]
@0xmovses is there a good way to test this e2e? Or, should we trust in the forge ~sim and reserve the rest for our time working on operational bits?

**Comment by 0xmovses** [comment]
@l-monninger e2e will happen in another PR after #690 goes in.

**Comment by l-monninger** [comment]
I know this is implemented for the [Bilbao Model](https://github.com/movementlabsxyz/MIP/pulls), but ideally we don't have to report the insurance fund balance and it can instead be queried on-chain. Even if it's just an account, as long as we know the account we can call `balanceOf` to determine how much it can ensure.

**Comment by 0xmovses** [comment]
@l-monninger sounds good.

File: protocol-units/bridge/contracts/src/AtomicBridgeCounterpartyMOVE.sol [hunk 4]
```diff
@@ -22,17 +23,26 @@ contract AtomicBridgeCounterpartyMOVE is IAtomicBridgeCounterpartyMOVE, OwnableU
...
 
-    function initialize(address _atomicBridgeInitiator, address owner, uint256 _timeLockDuration) public initializer {
+    // Initialize with initiator, RateLimiter, owner, and time lock duration
+    function initialize(
+        address _atomicBridgeInitiator,
+        address _rateLimiter,
+        address owner,
+        uint256 _timeLockDuration
+    ) public initializer {
         if (_atomicBridgeInitiator == address(0)) revert ZeroAddress();
```
Comment (franck44) [inline protocol-units/bridge/contracts/src/AtomicBridgeCounterpartyMOVE.sol:39]: This is input validation and should be written as:
```Solidity
require(_AtomicBridgeInitiator != address(0), ZeroAddress());
```

File: protocol-units/bridge/contracts/src/AtomicBridgeCounterpartyMOVE.sol [hunk 5]
```diff
@@ -22,17 +23,26 @@ contract AtomicBridgeCounterpartyMOVE is IAtomicBridgeCounterpartyMOVE, OwnableU
...
+        if (_rateLimiter == address(0)) revert ZeroAddress();
```
Comment (franck44) [inline protocol-units/bridge/contracts/src/AtomicBridgeCounterpartyMOVE.sol:40]: Input validation:
```Solidity
require (_rateLimiter != address(0), ZeroAddress());
```

File: protocol-units/bridge/contracts/src/AtomicBridgeCounterpartyMOVE.sol [hunk 6]
```diff
@@ -41,6 +51,11 @@ contract AtomicBridgeCounterpartyMOVE is IAtomicBridgeCounterpartyMOVE, OwnableU
         atomicBridgeInitiatorMOVE = AtomicBridgeInitiatorMOVE(_atomicBridgeInitiator);
     }
 
+    function setRateLimiter(address _rateLimiter) external onlyOwner {
+        if (_rateLimiter == address(0)) revert ZeroAddress();
```
Comment (franck44) [inline protocol-units/bridge/contracts/src/AtomicBridgeCounterpartyMOVE.sol:55]: Change to:

```Solidity
require (_rateLimiter != address(0), ZeroAddress());
```

File: protocol-units/bridge/contracts/src/AtomicBridgeInitiatorMOVE.sol [hunk 7]
```diff
@@ -35,23 +36,25 @@ contract AtomicBridgeInitiatorMOVE is IAtomicBridgeInitiatorMOVE, OwnableUpgrade
...
         uint256 _timeLockDuration,
-        uint256 _initialPoolBalance
+        uint256 _initialPoolBalance,
+        address _rateLimiter
     ) public initializer {
-        if (_moveToken == address(0)) {
-            revert ZeroAddress();
-        }
+        if (_moveToken == address(0)) revert ZeroAddress();
+        if (_rateLimiter == address(0)) revert ZeroAddress();
```
Comment (franck44) [inline protocol-units/bridge/contracts/src/AtomicBridgeInitiatorMOVE.sol:51]: Input validation:
```Solidity
require (_rateLimiter != address(0), ZeroAddress());
```

**Review by franck44 — CHANGES_REQUESTED** [review]
We may use `require` for input validation. 
We have already modified the code to use `requires` in another (pending) [PR](https://github.com/movementlabsxyz/movement/pull/852) so would be good to fix for consistency.
There are only a few instances to fix, so that should be an easy modification.

**Pull request closed without merging**